url5e6a51063b_nuigroup_hci_for_java_tgz-pJ8-nui_squirt_demo_DemoSquirtApplicationJ8/src/nui/squirt/component/Frame.java

```diff
@@ -0,0 +1,221 @@
+package nui.squirt.component;
+
+import java.awt.Color;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ListIterator;
+
+import nui.squirt.Component;
+import nui.squirt.Container;
+import nui.squirt.ControlPoint;
+import processing.core.PApplet;
+import processing.core.PVector;
+
+
+public class Frame extends Rectangle implements Container {
+	
+	private List<Component> components = new ArrayList<Component>();
+	
+	private ControlPoint[] cornerCPs = new ControlPoint[4];
+	private static final int UPPER_LEFT = 0;
+	private static final int UPPER_RIGHT = 1;
+	private static final int LOWER_RIGHT = 2;
+	private static final int LOWER_LEFT = 3;
+	
+	public Frame(float x, float y, float w, float h) {
+		super(x,y,w,h);
+		setFillColor(new Color(75, 75, 75, 220));
+		setStrokeColor(new Color(0, 0, 0, 0));
+	}
+
+	public Frame(float x, float y) {
+		this(x, y, 100, 100);
+	}
+
+	public void render(PApplet p) {
+		super.render(p);
+		
+//		p.rect(-getWidth()*0.475F, -getHeight()*0.475F, getWidth()*0.05F, getHeight()*0.05F);
+//		p.rect(getWidth()*0.475F, -getHeight()*0.475F, getWidth()*0.05F, getHeight()*0.05F);
+//		p.rect(getWidth()*0.475F, getHeight()*0.475F, getWidth()*0.05F, getHeight()*0.05F);
+//		p.rect(-getWidth()*0.475F, getHeight()*0.475F, getWidth()*0.05F, getHeight()*0.05F);
+
+		List<Component> l = new ArrayList<Component>(getComponents());
+		for (Component c: l) {
+			c.preRender(p);
+			c.render(p);
+			c.postRender(p);
+		}
+	}
+
+	public void add(Component c) {
+		getComponents().add(c);
+		c.setParent(this);
+	}
+
+	public void add(Component c, Object constraints) {
+		this.add(c);
+	}
+
+	public List<Component> getComponents() {
+		return components;
+	}
+
+	public void remove(Component c) {
+		getComponents().remove(c);
+		c.setParent(null);
+	}
+	
+	@Override
+	public boolean canAcceptMoreControlPoints() {
+		if (super.canAcceptMoreControlPoints()) {
+			return true;
+		}
+		else {
+			for (Component c: getComponents()) {
+				if (c.canAcceptMoreControlPoints())
+					return true;
+			}
+			return false;
+		}
+	}
+	
+	@Override
+	public boolean isUnderPoint(ControlPoint cp) {
+		if (super.isUnderPoint(cp)) {
+			return true;
+		}
+		else {
+			for (Component c: getComponents()) {
+				if (c.isUnderPoint(cp))
+					return true;
+			}
+			return false;
+		}
+	}
+	
+	@Override
+	public boolean offer(ControlPoint cp) {
+		ListIterator<Component> i = getComponents().listIterator(getComponents().size());
+		while (i.hasPrevious()) {
+			Component c = i.previous();
+			if (c.isUnderPoint(cp) && c.offer(cp)) {
+				i.remove();
+				getComponents().add(c);
+				return true;
+			}
+		}
+		if (super.canAcceptMoreControlPoints() && super.isUnderPoint(cp)) {
+			return super.offer(cp);
+		}
+		else return false;
+	}
+	
+	@Override
+	public void controlPointCreated(ControlPoint cp) {
+		PVector l = transformToLocalSpace(new PVector(cp.getX(), cp.getY()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;cp&#39; (index: 0 | type: ControlPoint). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other methods validating the parameter: &#39;controlPointDied&#39;, &#39;controlPointUpdated&#39;, &#39;offer&#39;. The same parameter type is validated here for example: url5e6a51063b_nuigroup_hci_for_java_tgz-pJ8-nui_squirt_demo_DemoSquirtApplicationJ8/src/nui/squirt/component/Circle.java:107. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).